pkg/oc/cli/cmd/image/mirror/mirror.go

```diff
@@ -0,0 +1,667 @@
+
+		// convert source tags to digests
+		for srcTag, pushTargets := range src.tags {
+			desc, err := srcRepo.Tags(ctx).Get(ctx, srcTag)
+			if err != nil {
+				tagErrs = append(tagErrs, retrieverError{src: src.ref, err: fmt.Errorf("unable to retrieve source image %s by tag: %v", src.ref, err)})
+				continue
+			}
+			srcDigest := desc.Digest
+			glog.V(3).Infof("Resolved source image %s:%s to %s\n", src.ref, srcTag, srcDigest)
+			src.mergeIntoDigests(srcDigest, pushTargets)
+		}
+
+		canonicalFrom := srcRepo.Named()
+
+		for srcDigestString, pushTargets := range src.digests {
+			// load the manifest
+			srcDigest := digest.Digest(srcDigestString)
+			// var contentDigest digest.Digest / client.ReturnContentDigest(&contentDigest),
+			srcManifest, err := manifests.Get(ctx, digest.Digest(srcDigest), schema2ManifestOnly)
+			if err != nil {
+				digestErrs = append(digestErrs, retrieverError{src: src.ref, err: fmt.Errorf("unable to retrieve source image %s manifest: %v", src.ref, err)})
+				continue
+			}
+
+			// filter or load manifest list as appropriate
+			srcManifests, srcDigest, err := processManifestList(ctx, srcDigest, srcManifest, manifests, src.ref, o.includeDescriptor)
+			if err != nil {
+				digestErrs = append(digestErrs, retrieverError{src: src.ref, err: err})
+				continue
+			}
+			if len(srcManifests) == 0 {
+				fmt.Fprintf(o.ErrOut, "info: Filtered all images from %s, skipping\n", src.ref)
+				continue
+			}
+
+			for _, dst := range pushTargets {
+				// if we are going to be using cross repository mount, get a token that covers the src
+				if src.ref.Registry == dst.ref.Registry {
+					toContext = toContext.WithScopes(auth.RepositoryScope{Repository: src.ref.RepositoryName(), Actions: []string{"pull"}})
+				}
+
+				toRepo, err := o.Repository(ctx, toContext, creds, dst.t, dst.ref)
+				if err != nil {
+					digestErrs = append(digestErrs, retrieverError{src: src.ref, dst: dst.ref, err: fmt.Errorf("unable to connect to %s: %v", dst.ref, err)})
+					continue
+				}
+
+				canonicalTo := toRepo.Named()
+				toManifests, err := toRepo.Manifests(ctx)
+				if err != nil {
+					digestErrs = append(digestErrs, retrieverError{src: src.ref, dst: dst.ref, err: fmt.Errorf("unable to access destination image %s manifests: %v", src.ref, err)})
+					continue
+				}
+
+				// if the destination tag already has this manifest, do nothing
+				var mustCopyLayers bool
+				if o.Force {
+					mustCopyLayers = true
+				} else {
+					if _, err := toManifests.Get(ctx, srcDigest); err != nil {
+						mustCopyLayers = true
+					} else {
+						glog.V(4).Infof("Manifest exists in %s, no need to copy layers without --force", dst.ref)
+					}
+				}
+				if mustCopyLayers {
+					// upload all the blobs
+					toBlobs := toRepo.Blobs(ctx)
+					srcBlobs := srcRepo.Blobs(ctx)
+
+					// upload the each manifest
+					for _, srcManifest := range srcManifests {
+						switch srcManifest.(type) {
+						case *schema2.DeserializedManifest:
```

Comment:
I don't really see a difference between handling schema2 and schema1.

Reply:
forcing function to stop propagation of schema1.  It's easy enough to add later.